pkg/odo/cli/utils/migrate.go

```diff
@@ -0,0 +1,349 @@
+	mo.componentName = mo.context.LocalConfigInfo.GetName()
+	return nil
+
+}
+
+// Validate validates the MigrateOptions based on completed values
+func (mo *MigrateOptions) Validate() (err error) {
+	if mo.context.LocalConfigInfo.GetSourceType() == config.GIT {
+		return errors.New("migration of git type s2i components to devfile not supported by odo")
+	}
+
+	return nil
+}
+
+// Run contains the logic for the command
+func (mo *MigrateOptions) Run() (err error) {
+
+	/*  This data is yet to be converted
+
+	// Absolute path
+	sourcePath, _ := context.LocalConfigInfo.GetOSSourcePath()
+	minMemory := context.LocalConfigInfo.GetMinMemory()
+	minCPU := context.LocalConfigInfo.GetMinCPU()
+	maxCPU := context.LocalConfigInfo.GetMaxCPU()
+
+	*/
+
+	err = generateDevfileYaml(mo)
+	if err != nil {
+		return errors.Wrap(err, "Error in generating devfile.yaml")
+	}
+
+	err = generateEnvYaml(mo)
+	if err != nil {
+		return errors.Wrap(err, "Error in generating env.yaml")
+	}
+
+	// TODO: Delete the s2i component and deploy the devfile component.
+
+	return nil
+}
+
+// NewCmdMigrate implements the odo utils migrate-to-devfile command
+func NewCmdMigrate(name, fullName string) *cobra.Command {
+	o := NewMigrateOptions()
+	migrateCmd := &cobra.Command{
+		Use:     name,
+		Short:   "migrates s2i based components to devfile based components",
+		Long:    migrateLongDesc,
+		Example: fmt.Sprintf(migrateExample, fullName),
+		Args:    cobra.ExactArgs(0),
+		Run: func(cmd *cobra.Command, args []string) {
+			genericclioptions.GenericRun(o, cmd, args)
+		},
+	}
+	return migrateCmd
+}
+
+func generateDevfileYaml(m *MigrateOptions) error {
+
+	// builder image to use
+	componentType := m.context.LocalConfigInfo.GetType()
+	// git, local, binary, none
+	sourceType := m.context.LocalConfigInfo.GetSourceType()
+
+	imageStream, imageforDevfile, err := getImageforDevfile(m.context.Client, componentType)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get image details")
+	}
+
+	envVarList := m.context.LocalConfigInfo.GetEnvVars()
+	s2iEnv, err := occlient.GetS2IEnvForDevfile(string(sourceType), envVarList, *imageStream)
+	if err != nil {
+		return err
+	}
+
+	s2iDevfile, err := data.NewDevfileData(devfileVersion)
+	if err != nil {
+		return err
+	}
+
+	// set schema version
+	s2iDevfile.SetSchemaVersion(devfileVersion)
+
+	// set metadata
+	s2iDevfile.SetMetadata(m.componentName, "1.0.0")
+	// set commponents
+	setDevfileComponentsForS2I(s2iDevfile, imageforDevfile, m.context.LocalConfigInfo, s2iEnv)
+	// set commands
+	setDevfileCommandsForS2I(s2iDevfile)
+
+	devObj := parser.DevfileObj{
+		Ctx:  devfileCtx.NewDevfileCtx(m.componentContext), //component context needs to be passed here
+		Data: s2iDevfile,
+	}
+
+	err = devObj.WriteYamlDevfile()
+	if err != nil {
+		return err
+	}
+	log.Italic("devfile.yaml is available in current directory, run `odo push` to deploy devfile component and `odo delete` to delete s2i component.\n")
```

Comment:
`delete s2i component` or devfile?

Reply:
you have an s2i component in folder
- cd my-s2i-component
- run `odo utils migrate-to-devfile` (generates devfile.yaml, env.yaml) 
- odo push (since devfile has priority over s2i, devfile component would get deployed. s2i would is still exisiting)
- odo delete(s2i has priority over for delete, so s2i would get deleted)
- you have now running devfile component.

Reply:
This sounds confusing TBH. How about we suggest delete before push? Will that work?

Reply:
yes, but what would happen when odo push for devfile gets failed and user lost the old s2i component as well?
in ideal migration strategy the when the new deployment gets successful then only the old one rolled back.

Reply:
OK :+1: Please add a comment regarding this.